internal/docs/design.md

```diff
@@ -374,6 +375,100 @@ b, err := blob.NewBucket(d, blob.FeatureUnicodeNames)
+
+These differences lead to a loss of portability and predictability for users.
+
+We considered restricting Go CDK's APIs to strings that all providers support.
+For example, we asserting that Go CDK's blob only supports ASCII plus `/` for
+blob names (and no `//`!). However, such a rule would mean that we couldn't
+handle existing strings that were created through some mechanism other than
+through Go CDK APIs. For example, an existing blob in S3 with a unicode name
+wouldn't be visible in Go CDK. Or maybe it would be visible, but you wouldn't be
+able to write to it. For metadata keys, rules like this might mean that you
+could read metadata attributes via the Go CDK and then write them back, and
+either lose data (if previously-existing invalid keys were dropped) or get an
+error. Overall, we decided that this approach is unacceptable.
+
+Instead, we assert that Go CDK can handle any string, and force provider
+implementations to handle strings that the underlying provider can't handle
+using escaping mechanisms. We enforce provider compliance with conformance
+tests. With this approach
+
+*   Go CDK APIs have visibility into all existing strings for all providers.
+    *   A slight caveat -- strings that look like Go CDK-escaped strings will be
+        unescaped by Go CDK. To mitigate this, we'll choose unusual-looking
+        escape mechanisms.
+*   Go CDK APIs are consistent in that you can write any string to any provider
+    and read it back.
+*   Some strings that were written through the Go CDK will appear in their
+    escaped form when viewed outside of Go CDK.
+
+The Go CDK providers helpers for such escaping:
+
+```
+package stringescape
+
+// Escaper supports escaping and unescaping of strings.
+type Escaper interface {
+    // Escape returns the string with unsupported characters escaped.
+    Escape(string) string
+    // Unescape reverses the escaping done by Escape.
+    Unescape(string) string
+}
+
+// NewNoopEscaper returns an Escaper that doesn't change the string at all.
+func NewNoopEscaper() Escaper { ... }
+
+// NewUnderscoreEscaper returns an Escaper that escapes each character for which
+// shouldEscape returns true with "__0xXX__", where XX is the hex representation
+// of the character's byte value.
+// shouldEscape takes the whole string and an index instead of a single
+// character because some escape decisions require context. For example, we
+// might want to escape the second "/" in "//" but not the first one.
+func NewUnderscoreEscaper(shouldEscape func(s string, i int) bool) Escaper {...}
+```
+
+To see how `Escaper` is used, let's use Azure's blob metadata keys as an
+example. As stated above, only valid `C#` identifiers are legal. The Azure blob
+provider would include code like this:
+
+```
+// shouldEscapeMetadataKey returns true if s[i] needs to be escaped in order
+// to produce a valid C# identifier.
+func shouldEscapeMetadataKey(s string, i int) {
+    return !isValidCSharpIdentifier(s[i])
```

Comment:
This is harder if C# allows non-Latin-8 characters in identifiers. You might need `utf8.DecodeRuneInString`.

Also, I'd call the function `isValidCSharpIdentifierByte`.

Reply:
Don't think it does. Anyway, that's an exercise for the driver :-).

Done on rename.